sdk/examples/xo_go/src/sawtooth_xo/handler/handler.go

```diff
@@ -0,0 +1,321 @@
+	}
+
+	if len(payload.Action) < 1 {
+		return nil, &processor.InvalidTransactionError{"Action is required"}
+	}
+
+	if payload.Action == "take" {
+		space, err := strconv.Atoi(parts[2])
+		payload.Space = space
+		if err != nil {
+			return nil, &processor.InvalidTransactionError{
+				fmt.Sprintf("Invalid Space : '%v'", parts[2])}
+		}
+	}
+
+	return &payload, nil
+}
+
+func unpackHeader(headerData []byte) (*transaction_pb2.TransactionHeader, error) {
+	header := &transaction_pb2.TransactionHeader{}
+	err := proto.Unmarshal(headerData, header)
+	if err != nil {
+		return nil, &processor.InternalError{
+			fmt.Sprint("Failed to unmarshal TransactionHeader: %v", err)}
+	}
+	return header, nil
+}
+
+func unpackGame(gameData []byte) (*Game, error) {
+	parts := strings.Split(string(gameData), ",")
+	if len(parts) != 5 {
+		return nil, &processor.InternalError{
+			fmt.Sprintf("Failed to deserialize game data: '%v'", string(gameData))}
+	}
+
+	game := &Game{
+		Board:   parts[0],
+		State:   parts[1],
+		Player1: parts[2],
+		Player2: parts[3],
+		Name:    parts[4],
+	}
+
+	switch game.State {
+	case "P1-WIN", "P2-WIN", "TIE", "P1-NEXT", "P2-NEXT":
+	default:
+		return nil, &processor.InternalError{
+			fmt.Sprintf("Game has reached invalid state: '%v'", game.State)}
+	}
+
+	return game, nil
+}
+
+func packGame(game *Game) []byte {
+	var buffer bytes.Buffer
+	buffer.WriteString(game.Board)
+	buffer.WriteString(",")
+	buffer.WriteString(game.State)
+	buffer.WriteString(",")
+	buffer.WriteString(game.Player1)
+	buffer.WriteString(",")
+	buffer.WriteString(game.Player2)
+	buffer.WriteString(",")
+	buffer.WriteString(game.Name)
+	return buffer.Bytes()
+}
+
+func loadGame(name string, state *processor.State) (*Game, error) {
+	// Use the namespace prefix + the hash of the game name to create the
+	// storage address
+	address := namespace + hexdigest(name)
+
+	results, err := state.Get([]string{address})
+	if err != nil {
+		return nil, &processor.InternalError{fmt.Sprint("Error getting state:", err)}
+	}
+
+	if len(string(results[address])) > 0 {
+		game, err := unpackGame(results[address])
+		if err != nil {
+			return nil, err
+		}
+
+		// NOTE: Since the game data is stored in a Merkle tree, there is a
+		// small chance of collision. A more correct usage would be to store
+		// a dictionary of games so that multiple games could be store at
+		// the same location. See the python intkey handler for an example
+		// of this.
+		if game.Name != name {
+			return nil, &processor.InternalError{"Hash collision"}
+		}
+		return game, nil
+	}
+	return nil, nil
+}
+
+func saveGame(game *Game, state *processor.State) error {
+	address := namespace + hexdigest(game.Name)
+	data := packGame(game)
+
+	addresses, err := state.Set(map[string][]byte{
+		address: data,
+	})
+	if err != nil {
+		return &processor.InternalError{fmt.Sprint("Failed to set new Value:", err)}
+	}
+	if len(addresses) == 0 {
+		return &processor.InternalError{"No addresses in set response"}
```

Comment:
Hmm, maybe I got the level of abstraction wrong with the state.Set method. Would it make more sense if the State class returned an error when the SetResponse does not include one of the addresses that it tried to set, even if the return status is okay?